### PR TITLE
Asynchronous secret validation

### PR DESCRIPTION
**Pull request by Phanatic**
### Description of change
Currently, the `secret.matches` function is synchronous, which causes the server to lockup while it reads the secret file from the file system before it continues execution. I've introduced a change to make the `secret.matches` function asynchronous and updated tests to match the async function.
This change also consolidates secret checking into a middleware function that can be reused in any new functions that need secret checking.

### Tests output
https://gist.github.com/Phanatic/9a9a6113413e39a7cd796d110a28470d

#####  1. Server running without a secret file.
*Client*
``` bash
curl -XPOST http://localhost:8080/flush;echo
{"error":"Missing or invalid secret"}
```

*Server*

``` bash
error: FATAL ERROR (Error: ENOENT: no such file or directory, open '/Users/phaniraj/Documents/github/ka/react-render-server/secret'): You must create a file:
error:     /Users/phaniraj/Documents/github/ka/react-render-server/secret
error: Its contents should be the secret-string at
error:     https://phabricator.khanacademy.org/K121
info: POST /flush 400 7ms
Error: Missing or invalid secret
    at renderSecret.matches (/Users/phaniraj/Documents/github/ka/react-render-server/src/server.js:115:20)
    at get (/Users/phaniraj/Documents/github/ka/react-render-server/src/secret.js:49:16)
    at ReadFileContext.fs.readFile [as callback] (/Users/phaniraj/Documents/github/ka/react-render-server/src/secret.js:33:14)
    at FSReqWrap.readFileAfterOpen [as oncomplete] (fs.js:365:13)
```
##### 2. Call `/render` without a secret.

*Client*
``` bash
curl -XPOST http://localhost:8080/render;echo
{"error":"Missing or invalid secret"}
```

*Server*
``` bash
info: react-render-server running at http://:::8080
info: POST /render 400 9ms
```

##### 3. Call `/flush` without a secret.

*Client*

 ``` bash
curl -XPOST http://localhost:8080/flush;echo
{"error":"Missing or invalid secret"}
```

*Server*

``` bash
info: POST /flush 400 1ms
```

### CODE REVIEW

File: src/secret.js [hunk 1]
```diff
@@ -20,30 +20,39 @@ const secretPath = path.normalize(__dirname + "/../secret");
...
-            logging.error('    ' + secretPath);
-            logging.error('Its contents should be the secret-string at');
-            logging.error('    https://phabricator.khanacademy.org/K121');
-            throw err;
-        }
+        return done(new Error('secret file is empty!'));
     }
-    return secret;
+
+    return done(secret);
```
Comment (csilvers) [inline src/secret.js]: Shouldn't this be `return done(null, secret)`?

If so, I'm concerned that testing didn't catch this.  You should probably write a test that exercises this code, and verifies it fails with the current code before you fix it.

File: src/secret.js [hunk 2]
```diff
@@ -20,30 +20,39 @@ const secretPath = path.normalize(__dirname + "/../secret");
...
+  });
 };
 
 
-const matches = function(actual) {
-    return get() === actual;
+const matches = function(actual, done) {
```
Comment (csilvers) [inline src/secret.js]: I don't understand the API of `matches` anymore.  It calls `done` with what parameter?  Either a bool or an exception?  That seems weird.  At the least I'd expect it to follow the `(err, retval)` pattern you do above.  You should document the semantics here, and possibly the reason behind them.

File: src/secret.js [hunk 3]
```diff
@@ -20,30 +20,39 @@ const secretPath = path.normalize(__dirname + "/../secret");
 let secret;
 
 // Used by the benchmark/loadtest tool.
-const get = function() {
+const get = function(done) {
+  if (secret) {
+    return done(null, done);
```
Comment (Phanatic) [inline src/secret.js]: You're right! I missed that in my tests. I'll fix the code and add a test for this.

File: src/secret.js [hunk 4]
```diff
@@ -20,30 +20,39 @@ const secretPath = path.normalize(__dirname + "/../secret");
...
+  }
+  fs.readFile(secretPath, "utf-8", (err, contents) =>{
```
Comment (Phanatic) [inline src/secret.js]: Yep, will do. Am curious if there is a style guide I should follow?

File: src/secret.js (hunk 1 above)
Comment (Phanatic) [inline src/secret.js]: Yeah, this was another thing I noticed on my second run through the code. I've added a separate test for the `secret` module so we can catch this stuff in the future.

File: src/secret.js (hunk 2 above)
Comment (Phanatic) [inline src/secret.js]: Yep, I missed passing the error value out of here.
The expectation for the caller is that the callback is called with `(error, matchFound)`.
Fixed it now.

**Comment by csilvers** [comment]
Looks great!  Thanks for doing this.

We have a js style guide at https://github.com/Khan/style-guides/blob/master/style/javascript.md

**Pull request merged**